### DISCUSSION TITLE
YUV to RGBA texture conversion using material shader?

### DISCUSSION THREAD

**Discussion by imbrig** [google/filament#4627] [Q&A] [answered]
I have an gltf asset that uses a video file as a texture. The video is controlled from outside of Filament and outputs each frame as android.media.Image in YUV format which I can pass through the JNI as byte arrays to create PixelBufferDescriptors. I was thinking about using a Filament shader / material to do the conversion to RGBA and passing the results to the gltf material attached to the mesh. I can't replace the gltfio generated material already assigned to the mesh due to overwriting material settings from the gltf.

I was wondering is there's a way to chain the pixel shaders together? So that one renders the texture and passes it off to the next? Another solution would be to render the yuv to rgba shader to a texture which is just applied normally to the mesh, or anything else? Thanks.

**Comment by romainguy** [comment, marked as answer]
Can you use an external texture instead? They're perfect for exactly this use case: play the video in a `SurfaceTexture`, and use that in Filament. It's the most efficient way.

**Reply by imbrig** [reply]
Most of the cases I do use `SurfaceTexture` with `Stream`, and it works perfectly. Just this one use case where I do not have a surface but receive frame updates of `android.media.Image`. Currently I can convert the YUV Image to an rgba byte array / byte buffer using a`RenderScript`, or I can have Filament do the conversion in a shader for optimization. The trick for me was to not overwrite the original material created in gltfio or it would just be a matter of applying the YUV to rgba material.

Maybe the question I should ask would be if there's a way to preprocess a texture with another pixel shader? One solution would be creating a render target texture that renders that off screen. I remember a sample with a statue that did a render to quad (maybe it was removed), but instead of rendering a view or scene it would just be combining a texture.

**Reply by romainguy** [reply]
> One solution would be creating a render target texture that renders that off screen

Yes, that's the right way to do this. There's a [C++ sample](https://github.com/google/filament/blob/main/samples/rendertarget.cpp#L213) for this.

**Reply by imbrig** [reply]
Thanks! I'll try using that solution.
